src/Components/Blazor/Blazor/src/Hosting/WebAssemblyHostBuilder.cs

```diff
@@ -72,6 +79,30 @@ private WebAssemblyHostBuilder()
         /// </summary>
         public IServiceCollection Services { get; }
 
+        /// <summary>
+        /// Registers a <see cref="IServiceProviderFactory{TBuilder}" /> instance to be used to create the <see cref="IServiceProvider" />.
+        /// </summary>
+        /// <param name="factory">The <see cref="IServiceProviderFactory{TBuilder}" />.</param>
+        /// <param name="configure">
+        /// A delegate used to configure the <typeparamref T="TBuilder" />. This can be used to configure services using
+        /// APIS specific to the <see cref="IServiceProviderFactory{TBuilder}" /> implementation.
+        /// </param>
+        /// <typeparam name="TBuilder">The type of builder provided by the <see cref="IServiceProviderFactory{TBuilder}" />.</typeparam>
+        public void ConfigureContainer<TBuilder>(IServiceProviderFactory<TBuilder> factory, Action<TBuilder> configure)
+        {
+            if (factory == null)
+            {
+                throw new ArgumentNullException(nameof(factory));
+            }
+
+            _createServiceProvider = () =>
+            {
+                var container = factory.CreateBuilder(Services);
+                configure?.Invoke(configure);
+                return factory.CreateServiceProvider(container);
+            }
+        }
+
```

Comment:
I approached this with two goals (in order):

- Use `IServiceProviderFactory<>` since that's what everyone integrates with already
- Make this simple (avoid callbacks)

---

Unfortunately, since `IServiceProviderFactory<>` is generic, there aren't great ways to avoid callbacks without making **all** of this stuff generic (yuck). That seemed like a really bad change to satisfy the ability to use different containers.

So since I couldn't avoid a callback, I did as much as I thought possible to make it easy to use.

---

Code Samples:

```C#
var builder = WebAssemblyHostBuilder.CreateDefault(args);
builder.ConfigureContainer(new AutoFacServiceProviderFactory());

...
await builder.Build().RunAsync();
```

```C#
var builder = WebAssemblyHostBuilder.CreateDefault(args);
builder.ConfigureContainer(new AutoFacServiceProviderFactory(), () =>
{
    // Use AutoFac APIs
});

...
await builder.Build().RunAsync();
```

Reply:
If you care about supporting existing extension methods you'll need the old overload as well. I came across the same problem here https://github.com/davidfowl/FeatherHttp#uber-example.

Reply:
It really seems to me like making this a single method to register the provider and add a callback is much simpler than separate methods, because you don't have to know the generic type to fill in.

Reply:
@davidfowl can you be more specific? We don't implement `IHostBuilder` so I'm not sure what existing methods you mean.

Reply:
Also did you try to simplify this in feather? I looked for a place where you did and didn't see anything.

Reply:
> @davidfowl can you be more specific? We don't implement IHostBuilder so I'm not sure what existing methods you mean.

Ah, then you're fine. I tried to do the single method thing but it blocks existing extension methods on IHostBuilder (which I wanted to support). Some of the IServiceProviderFactory implementations didn't have publicly exposed implementations of IServiceProviderFactory and used the extension method for registration.

Reply:
> Some of the IServiceProviderFactory implementations didn't have publicly exposed implementations of IServiceProviderFactory and used the extension method for registration.

That gives me a little bit of pause then. There are two ways to go:

1. Decide that everyone has to make an `IServiceProviderFactory<>` 
2. We try to create something simpler

Reply:
I'm pretty sure AutoFac has an exposed implementation right? AutoFac is what people said they were trying to use.

Reply:
Autofac does have a public impl.

Reply:
I tried making something simpler here and it has problems. Basically because we need the service provider factory to run last, there's no real way to do it other than a callback. And if you're making a callback, you want something just like what we have. 

Let's go with this, and we can add more options in the future if we think of something better.